--- a/net/Kconfig
+++ b/net/Kconfig
@@ -460,6 +460,18 @@ config FAILOVER
 	  migration of VMs with direct attached VFs by failing over to the
 	  paravirtual datapath when the VF is unplugged.
 
+config HW_NAT
+	bool "HW NAT support"
+	default n
+	---help---
+	 This feature provides a fast path to support network lan/wan nat.
+	 If you need hw_nat engine to reduce cpu loading, please say Y.
+
+	  Note that the answer to this question doesn't directly affect the
+	  kernel: saying N will just cause the configurator to skip all
+	  the questions about Mediatek Ethernet devices. If you say Y,
+	  you will be asked for your specific card in the following questions.
+
 endif   # if NET
 
 # Used by archs to tell that they support BPF JIT compiler plus which flavour.
--- a/net/Makefile
+++ b/net/Makefile
@@ -62,6 +62,9 @@ endif
 obj-$(CONFIG_6LOWPAN)		+= 6lowpan/
 obj-$(CONFIG_IEEE802154)	+= ieee802154/
 obj-$(CONFIG_MAC802154)		+= mac802154/
+ifeq ($(CONFIG_HW_NAT),y)
+obj-y                           += nat/foe_hook/
+endif
 
 ifeq ($(CONFIG_NET),y)
 obj-$(CONFIG_SYSCTL)		+= sysctl_net.o
