--- a/iwinfo_utils.c
+++ b/iwinfo_utils.c
@@ -213,6 +213,16 @@ int iwinfo_hardware_id_from_mtd(struct i
 		if (ident == 0x7981 || ident == 0x7986
 			|| ident == 0x8179 || ident == 0x8679)
 		{
+			uint16_t ident_zone2 = 0;
+			if (len > 0xa0002 && 
+				lseek(fd, 0xa0000, SEEK_SET) != -1 &&
+				read(fd, &ident_zone2, sizeof(ident_zone2)) != -1)
+			{
+				if (ident_zone2 != 0xFFFF && ident_zone2 != 0x0) {
+					close(fd);
+					return -1;
+				}
+			}
 			if ((ident & 0xff) == 0x79)
 				id->device_id = (ident >> 8) | (ident & 0x00ff) << 8;
 			else
@@ -222,6 +232,10 @@ int iwinfo_hardware_id_from_mtd(struct i
 			id->subsystem_device_id = id->device_id;
 			close(fd);
 			return 0;
+		} else if (ident == 0xffff || ident == 0x0) {
+			/* eeprom invalid */
+			close(fd);
+			return -1;
 		}
 	}
 	lseek(fd, 0, SEEK_SET);
